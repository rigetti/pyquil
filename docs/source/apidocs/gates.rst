Gates and Instructions
======================

A :py:class:`~pyquil.quil.Program` is effectively a list of gates and instructions which
can be created using the function documented in this section::

    >>> p = Program()
    >>> p += H(qubit=0)
    >>> p += RY(angle=pi/3, qubit=1)
    >>> p += CNOT(0, 1)
        ...


.. currentmodule:: pyquil.gates

Native gates for Rigetti QPUs
-----------------------------

Physical quantum processors can enact a subset of all named gates. Luckily,
a small set of gates is universal for quantum computation, so all named gates
can be enacted by suitable combinations of physically realizable gates. Rigetti's
superconducting quantum processors can perform :py:func:`RX` with ``angle=+-pi/2`` or
``angle=+-pi``, :py:func:`RZ` with an arbitrary angle, and :py:func:`CZ` interactions
between neighboring qubits. Rigetti QPUs can natively measure in the computational (Z) basis.

.. autosummary::
    :toctree: autogen
    :template: autosumm.rst

    RX
    RZ
    CZ
    MEASURE


All gates and instructions
--------------------------

In general, you will write a quantum program using the full suite of Quil gates and instructions
and use :ref:`the Quil compiler <compiler>` to compile programs into the target instruction
set (ISA). The full list of quantum gates and classical Quil instructions is enumerated here.

.. rubric:: Single-qubit gates
.. autosummary::
    :toctree: autogen
    :template: autosumm.rst

    I
    X
    Y
    Z
    H
    S
    T
    RX
    RY
    RZ
    PHASE

.. rubric:: Multi-qubit gates
.. autosummary::
    :toctree: autogen
    :template: autosumm.rst

    CZ
    CNOT
    CCNOT
    CPHASE00
    CPHASE01
    CPHASE10
    CPHASE
    SWAP
    CSWAP
    ISWAP
    PSWAP


.. rubric:: Classical instructions
.. autosummary::
    :toctree: autogen
    :template: autosumm.rst

    WAIT
    RESET
    NOP
    HALT
    MEASURE
    NEG
    NOT
    AND
    OR
    IOR
    XOR
    MOVE
    EXCHANGE
    LOAD
    STORE
    CONVERT
    ADD
    SUB
    MUL
    DIV
    EQ
    LT
    LE
    GT
    GE

.. rubric:: Collections
.. autosummary::

    QUANTUM_GATES
    STANDARD_INSTRUCTIONS

