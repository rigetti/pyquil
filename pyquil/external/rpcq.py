from typing import Dict, List, Union, Optional, Any
from typing_extensions import Literal
from pydantic import BaseModel, Field
from rpcq.messages import TargetDevice as TargetQuantumProcessor


class Operator(BaseModel):
    operator: Optional[str] = None
    duration: Optional[float] = None
    fidelity: Optional[float] = None


class MeasureInfo(Operator):
    qubit: Optional[Union[int, str]] = None
    target: Optional[Union[int, str]] = None
    operator_type: Literal["measure"] = "measure"


class GateInfo(Operator):
    parameters: List[Union[float, str]] = Field(default_factory=list)
    arguments: List[Union[int, str]] = Field(default_factory=list)
    operator_type: Literal["gate"] = "gate"


class Qubit(BaseModel):
    id: int
    dead: Optional[bool] = False
    gates: List[Union[GateInfo, MeasureInfo]] = Field(default_factory=list)

    def dict(self, **kwargs: Any) -> Dict[str, Any]:
        exclude = kwargs.get("exclude") or set()
        if not self.dead:
            exclude.add("dead")
        kwargs["exclude"] = exclude
        return super().dict(**kwargs)


class Edge(BaseModel):
    ids: List[int]
    dead: Optional[bool] = False
    gates: List[GateInfo] = Field(default_factory=list)

    def dict(self, **kwargs: Any) -> Dict[str, Any]:
        exclude = kwargs.get("exclude") or set()
        if not self.dead:
            exclude.add("dead")
        kwargs["exclude"] = exclude
        return super().dict(**kwargs)


class CompilerISA(BaseModel):
    qubits: Dict[str, Qubit] = Field(default_factory=dict, alias="1Q")
    edges: Dict[str, Edge] = Field(default_factory=dict, alias="2Q")


def add_qubit(quantum_processor: CompilerISA, node_id: int) -> Qubit:
    if node_id not in quantum_processor.qubits:
        quantum_processor.qubits[str(node_id)] = Qubit(id=node_id)
    return quantum_processor.qubits[str(node_id)]


def get_qubit(quantum_processor: CompilerISA, node_id: int) -> Optional[Qubit]:
    return quantum_processor.qubits.get(str(node_id))


def make_edge_id(qubit1: int, qubit2: int) -> str:
    return "-".join([str(qubit) for qubit in sorted([qubit1, qubit2])])


def add_edge(quantum_processor: CompilerISA, qubit1: int, qubit2: int) -> Edge:
    edge_id = make_edge_id(qubit1, qubit2)
    if edge_id not in quantum_processor.edges:
        quantum_processor.edges[edge_id] = Edge(ids=sorted([qubit1, qubit2]))
    return quantum_processor.edges[edge_id]


def get_edge(quantum_processor: CompilerISA, qubit1: int, qubit2: int) -> Optional[Edge]:
    edge_id = make_edge_id(qubit1, qubit2)
    return quantum_processor.edges.get(edge_id)


def _edge_ids_from_id(edge_id: str) -> List[int]:
    return [int(node_id) for node_id in edge_id.split("-")]


def _compiler_isa_from_dict(data: Dict[str, Dict[str, Any]]) -> CompilerISA:
    compiler_isa_data = {
        "1Q": {k: {"id": int(k), **v} for k, v in data.get("1Q", {}).items()},
        "2Q": {k: {"ids": _edge_ids_from_id(k), **v} for k, v in data.get("2Q", {}).items()},
    }
    return CompilerISA.parse_obj(compiler_isa_data)


def compiler_isa_to_target_quantum_processor(compiler_isa: CompilerISA) -> TargetQuantumProcessor:
    return TargetQuantumProcessor(isa=compiler_isa.dict(by_alias=True), specs={})


class Supported1QGate:
    I = "I"
    RX = "RX"
    RZ = "RZ"
    MEASURE = "MEASURE"
    WILDCARD = "WILDCARD"


class Supported2QGate:
    WILDCARD = "WILDCARD"
    CZ = "CZ"
    ISWAP = "ISWAP"
    CPHASE = "CPHASE"
    XY = "XY"
