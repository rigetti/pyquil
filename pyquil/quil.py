#!/usr/bin/python
##############################################################################
# Copyright 2016-2017 Rigetti Computing
#
#    Licensed under the Apache License, Version 2.0 (the "License");
#    you may not use this file except in compliance with the License.
#    You may obtain a copy of the License at
#
#        http://www.apache.org/licenses/LICENSE-2.0
#
#    Unless required by applicable law or agreed to in writing, software
#    distributed under the License is distributed on an "AS IS" BASIS,
#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#    See the License for the specific language governing permissions and
#    limitations under the License.
##############################################################################
"""
Module for creating and defining Quil programs.
"""

from .quilbase import (InstructionGroup,
                       Addr,
                       While,
                       If,
                       DefGate,
                       Gate,
                       Measurement,
                       merge_resource_managers)

from .gates import MEASURE, STANDARD_GATES
from math import pi


class Program(InstructionGroup):
    def __init__(self, *instructions):
        super(Program, self).__init__()
        self.inst(*instructions)
        self.defined_gates = []

    def synthesize(self, resource_manager=None):
        self.resource_manager.reset()
        return super(Program, self).synthesize(resource_manager)

    def __add__(self, other):
        """
        Concatenate two programs together, returning a new one.

        :param Program other: Another program or instruction to concatenate to this one.
        :return: A newly concatenated program.
        :rtype: Program
        """
        if isinstance(other, Program):
            p = Program()
            p.defined_gates = self.defined_gates + other.defined_gates
            p.actions = self.actions + other.actions
            p.resource_manager = merge_resource_managers(self.resource_manager,
                                                         other.resource_manager)
            return p
        else:
            return super(Program, self).__add__(other)

    def __getitem__(self, index):
        """
        Allows indexing into the program to get an action.
        :param index: The action at the specified index.
        :return:
        """
        return self.actions[index]

    def __iter__(self):
        """
        Allow built in iteration through a program's actions, e.g. [a for a in Program(X(0))]
        :return:
        """
        return self.actions.__iter__()

    def get_qubits(self):
        """
        :return: a set of all the qubit indices allocated in this program, synthesizing freely
        allocated qubits if neccessary.
        :rtype: set
        """
        allocated_qubits = set()
        self.synthesize()
        for ii, action in self:
            if isinstance(action, Gate):
                qubit_indices = {qq.index() for qq in action.arguments}
            elif isinstance(action, Measurement):
                qubit_indices = {action.arguments[0].index()}
            else:
                continue
            allocated_qubits = set.union(allocated_qubits, qubit_indices)
        return allocated_qubits

    def defgate(self, name, matrix):
        """
        Define a new static gate.

        :param string name: The name of the gate.
        :param array-like matrix: List of lists or Numpy 2d array.
        :return: The Program instance.
        :rtype: Program
        """
        self.defined_gates.append(DefGate(name, matrix))
        return self

    def measure(self, qubit_index, classical_reg):
        """
        Measures a qubit at qubit_index and puts the result in classical_reg

        :param int qubit_index: The address of the qubit to measure.
        :param int classical_reg: The address of the classical bit to store the result.
        :returns: The Quil Program with the appropriate measure instruction appended, e.g.
                  MEASURE 0 [1]
        :rtype: Program
        """
        return self.inst(MEASURE(qubit_index, classical_reg))

    def measure_all(self, *qubit_reg_pairs):
        """
        Measures many qubits into their specified classical bits, in the order
        they were entered.

        :param Tuple qubit_reg_pairs: Tuples of qubit indices paired with classical bits.
        :return: The Quil Program with the appropriate measure instructions appended, e.g.
                  MEASURE 0 [1]
                  MEASURE 1 [2]
                  MEASURE 2 [3]
        :rtype: Program
        """
        for qubit_index, classical_reg in qubit_reg_pairs:
            self.inst(MEASURE(qubit_index, classical_reg))
        return self

    def while_do(self, classical_reg, q_program):
        """
        While a classical register at index classical_reg is 1, loop q_program

        :param int classical_reg: The classical register to check
        :param Program q_program: The Quil program to loop.
        :return: The Quil Program with the loop instructions added.
        :rtype: Program
        """
        w_loop = While(Addr(classical_reg))
        w_loop.Body.inst(q_program)
        return self.inst(w_loop)

    def if_then(self, classical_reg, if_program, else_program=None):
        """
        If the classical register at index classical reg is 1, run if_program, else run
        else_program.

        :param int classical_reg: The classical register to check as the condition
        :param Program if_program: A Quil program to execute if classical_reg is 1
        :param Program else_program: A Quil program to execute if classical_reg is 0. This
            argument is optional and defaults to an empty Program.
        :returns: The Quil Program with the branching instructions added.
        :rtype: Program
        """

        else_program = else_program if else_program is not None else Program()

        branch = If(Addr(classical_reg))
        branch.Then.inst(if_program)
        branch.Else.inst(else_program)
        return self.inst(branch)

    def dagger(self, inv_dict=None, suffix="-INV"):
        """
        Creates the conjugate transpose of the Quil program. The program must not
        contain any irreversible actions (measurement, control flow, qubit allocation).

        :return: The Quil program's inverse
        :rtype: Program

        """

        for action in self.actions:
            assert action[0] == 0, "Program must be valid Protoquil"
            gate = action[1]
            assert not isinstance(gate, Measurement), "Program cannot contain measurements"
            assert not isinstance(gate, While) and not isinstance(gate, If), \
                "Program cannot contain control flow"

        daggered = Program()

        for gate in self.defined_gates:
            if inv_dict == None or gate.name not in inv_dict:
                daggered.defgate(gate.name + suffix, gate.matrix.T.conj())

        for action in self.actions[::-1]:
            gate = action[1]
            if gate.operator_name in STANDARD_GATES:
                if gate.operator_name == "S":
                    daggered.inst(STANDARD_GATES["PHASE"](-pi / 2, *gate.arguments))
                elif gate.operator_name == "T":
                    daggered.inst(STANDARD_GATES["RZ"](pi / 4, *gate.arguments))
                elif gate.operator_name == "ISWAP":
                    daggered.inst(STANDARD_GATES["PSWAP"](pi / 2, *gate.arguments))
                else:
                    negated_params = list(map(lambda x : -1*x, gate.parameters))
                    daggered.inst(STANDARD_GATES[gate.operator_name](*(negated_params+gate.arguments)))
            else:
                if inv_dict == None or gate.operator_name not in inv_dict:
                    gate_inv_name = gate.operator_name + suffix
                else:
                    gate_inv_name = inv_dict[gate.operator_name]

                daggered.inst(tuple([gate_inv_name] + gate.arguments))

        return daggered

    def out(self):
        """
        Converts the Quil program to a readable string.

        :return: String form of a program
        :rtype: string
        """
        s = ""
        for dg in self.defined_gates:
            s += dg.out()
            s += "\n"
        s += super(Program, self).out()
        return s

def merge_programs(prog_list):
    """
    Merges a list of pyQuil programs into a single one by appending them in sequence

    :param list prog_list: A list of pyquil programs
    :return: a single pyQuil program
    :rtype: Program
    """
    return sum(prog_list, Program())
