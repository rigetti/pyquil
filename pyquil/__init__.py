from pyquil._version import pyquil_version
from pyquil.quil import Program
from pyquil.api import list_quantum_computers, get_qc

__version__ = pyquil_version
