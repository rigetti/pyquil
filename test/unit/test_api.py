#!/usr/bin/python
##############################################################################
# Copyright 2016-2017 Rigetti Computing
#
#    Licensed under the Apache License, Version 2.0 (the "License");
#    you may not use this file except in compliance with the License.
#    You may obtain a copy of the License at
#
#        http://www.apache.org/licenses/LICENSE-2.0
#
#    Unless required by applicable law or agreed to in writing, software
#    distributed under the License is distributed on an "AS IS" BASIS,
#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#    See the License for the specific language governing permissions and
#    limitations under the License.
##############################################################################
from math import pi

import numpy as np

from pyquil.external.rpcq import CompilerISA, Edge, Qubit
from pyquil.gates import CNOT, CZ, MEASURE, PHASE, RX, RZ, H
from pyquil.paulis import PauliTerm
from pyquil.quil import Program
from pyquil.quilatom import MemoryReference
from pyquil.quilbase import Declare, Halt
from pyquil.simulation.tools import program_unitary

EMPTY_PROGRAM = Program()
BELL_STATE = Program(H(0), CNOT(0, 1))
BELL_STATE_MEASURE = Program(
    Declare("ro", "BIT", 2),
    H(0),
    CNOT(0, 1),
    MEASURE(0, MemoryReference("ro", 0)),
    MEASURE(1, MemoryReference("ro", 1)),
)
COMPILED_BELL_STATE = Program(
    [
        RZ(pi / 2, 0),
        RX(pi / 2, 0),
        RZ(-pi / 2, 1),
        RX(pi / 2, 1),
        CZ(1, 0),
        RZ(-pi / 2, 0),
        RX(-pi / 2, 1),
        RZ(pi / 2, 1),
        Halt(),
    ]
)
DUMMY_ISA = CompilerISA(
    qubits={
        "0": Qubit(id=0),
        "1": Qubit(id=1),
    },
    edges={
        "0-1": Edge(ids=[0, 1]),
    },
)

COMPILED_BYTES_ARRAY = b"SUPER SECRET PACKAGE"
RB_ENCODED_REPLY = [[0, 0], [1, 1]]
RB_REPLY = [Program("H 0\nH 0\n"), Program("PHASE(pi/2) 0\nPHASE(pi/2) 0\n")]


def test_quil_to_native_quil(compiler):
    response = compiler.quil_to_native_quil(BELL_STATE)
    p_unitary = program_unitary(response, n_qubits=2)
    compiled_p_unitary = program_unitary(COMPILED_BELL_STATE, n_qubits=2)
    from pyquil.simulation.tools import scale_out_phase

    assert np.allclose(p_unitary, scale_out_phase(compiled_p_unitary, p_unitary))


def test_local_rb_sequence(benchmarker, snapshot):
    response = benchmarker.generate_rb_sequence(2, [PHASE(np.pi / 2, 0), H(0)], seed=52)
    assert [prog.out() for prog in response] == snapshot


def test_local_conjugate_request(benchmarker):
    response = benchmarker.apply_clifford_to_pauli(Program("H 0"), PauliTerm("X", 0, 1.0))
    assert isinstance(response, PauliTerm)
    assert str(response) == "(1+0j)*Z0"


def test_apply_clifford_to_pauli(benchmarker):
    response = benchmarker.apply_clifford_to_pauli(Program("H 0"), PauliTerm("I", 0, 0.34))
    assert response == PauliTerm("I", 0, 0.34)
