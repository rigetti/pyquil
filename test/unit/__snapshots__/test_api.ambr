# serializer version: 1
# name: test_local_rb_sequence
  list([
    '''
      H 0
      PHASE(1.5707963267948966) 0
      H 0
      PHASE(1.5707963267948966) 0
      PHASE(1.5707963267948966) 0
  
    ''',
    '''
      H 0
      PHASE(1.5707963267948966) 0
      H 0
      PHASE(1.5707963267948966) 0
      PHASE(1.5707963267948966) 0
  
    ''',
  ])
# ---
